private/buf/bufgen/features.go

```diff
@@ -0,0 +1,121 @@
+	"google.golang.org/protobuf/types/pluginpb"
+)
+
+// requiredFeatures maps a feature to the set of files in an image that
+// make use of that feature.
+type requiredFeatures map[pluginpb.CodeGeneratorResponse_Feature][]string
+
+type featureChecker func(options *descriptorpb.FileDescriptorProto) bool
+
+// Map of all known features to functions that can check whether a given file
+// uses said feature.
+var allFeatures = map[pluginpb.CodeGeneratorResponse_Feature]featureChecker{
+	pluginpb.CodeGeneratorResponse_FEATURE_PROTO3_OPTIONAL: fileHasProto3Optional,
+}
+
+// computeRequiredFeatures returns a map of required features to the files in
+// the image that require that feature. After plugins are invoked, the plugins'
+// responses are checked to make sure any required features were supported.
+func computeRequiredFeatures(img bufimage.Image) requiredFeatures {
+	features := requiredFeatures{}
+	for feature, checker := range allFeatures {
+		for _, file := range img.Files() {
+			if file.IsImport() {
+				// we only want to check the sources in the module, not their dependencies
+				continue
+			}
+			if checker(file.Proto()) {
+				features[feature] = append(features[feature], file.Path())
+			}
+		}
+	}
+	return features
+}
+
+func checkRequiredFeatures(req requiredFeatures, resps []*pluginpb.CodeGeneratorResponse) {
+	for _, resp := range resps {
+		if resp == nil || resp.GetError() != "" {
+			// plugin failed, nothing to check
+			continue
+		}
+		failed := requiredFeatures{}
+		var failedFeatures []pluginpb.CodeGeneratorResponse_Feature
+		supported := resp.GetSupportedFeatures() // bit mask of features the plugin supports
+		for feature, files := range req {
+			featureMask := (uint64)(feature)
+			if supported&featureMask != featureMask {
+				// doh! Supported features don't include this one
+				failed[feature] = files
+				failedFeatures = append(failedFeatures, feature)
+			}
+		}
+		if len(failed) > 0 {
+			var buf bytes.Buffer
+			buf.WriteString("Plugin does not support required features.\n")
+			sort.Slice(failedFeatures, func(i, j int) bool {
+				return failedFeatures[i].Number() < failedFeatures[j].Number()
+			})
+			for _, feature := range failedFeatures {
+				files := failed[feature]
+				// bytes.Buffer does not generate I/O errors, so no need to check err result
+				_, _ = fmt.Fprintf(&buf, "%v, required by %d file(s):\n  ", feature, len(files))
+				_, _ = fmt.Fprintln(&buf, strings.Join(files, ","))
+			}
+			// clear out code gen results and replace with an error
+			resp.File = nil
+			resp.Error = proto.String(buf.String())
+		}
+	}
+}
+
+func fileHasProto3Optional(fd *descriptorpb.FileDescriptorProto) bool {
+	if fd.GetSyntax() != "proto3" {
+		// can't have proto3 optional unless syntax is proto3
+		return false
+	}
+	for _, msg := range fd.MessageType {
+		if msgHasProto3Optional(msg) {
+			return true
+		}
+	}
+	return false
+}
+
+func msgHasProto3Optional(md *descriptorpb.DescriptorProto) bool {
```

Comment:
`messageHasProto3Optional` and `descriptorProto`

Reply:
done